ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -234,7 +236,8 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOExceptio
           }
         }
 
-        JavaUtils.allOf(remoteWrites).thenCombine(localWrite, (v, bytesWritten) -> {
+        previous = previous.thenCombine(JavaUtils.allOf(remoteWrites), (u, v) -> null)
+            .thenCombine(localWrite, (v, bytesWritten) -> {
```

Comment:
this makes sense!

Reply:
@szetszwo Because the slow previous will slow down the following request. So I am considering another fix: Use Map instead of queue for the [replies](https://github.com/apache/incubator-ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java#L46). If we use queue, when send request1, request2, but response reply2, reply1, when client receive reply2 it will complete request1, which should complete request2,  then error happen. If we use Map instead of queue, we do not need to wait the previous.

Reply:
> ... slow previous will slow down the following request.  ...

It is a correct behaviour for Streaming.  Streaming is FIFO like a queue.  We should guarantee the replies are in-ordered.

It is not an unordered RPC which allows later requests replied before the earlier requests.